src/python/pants/bin/daemon_pants_runner.py

```diff
@@ -36,20 +36,25 @@ class DaemonExiter(Exiter):
 
   @classmethod
   @contextmanager
-  def override_global_exiter(cls, maybe_shutdown_socket: MaybeShutdownSocket, finalizer: Callable[[], None]) -> None:
-    with ExceptionSink.exiter_as(lambda previous_exiter: cls(maybe_shutdown_socket, finalizer, previous_exiter)):
+  def override_global_exiter(
+    cls, maybe_shutdown_socket: MaybeShutdownSocket, finalizer: Callable[[], None],
+  ) -> Iterator[None]:
+    with ExceptionSink.exiter_as(
+      lambda previous_exiter: cls(maybe_shutdown_socket, finalizer, previous_exiter)
+    ):
       yield
 
-  def __init__(self,
+  def __init__(
+    self,
     maybe_shutdown_socket: MaybeShutdownSocket,
     finalizer: Callable[[], None],
-    previous_exiter: Exiter):
-
-    super().__init__(exiter=previous_exiter)
+    previous_exiter: Optional[Exiter],
+  ) -> None:
+    super().__init__(exiter=previous_exiter or sys.exit)
```

Comment:
Where was the (secretly) `Optional[Exiter]` implicitly being converted into `sys.exit` before this change?

Reply:
Good question. This is all because we define the default exiter in ExceptionSink as `None` rather than `Exiter(sys.exit)`, which only gets set here:

https://github.com/pantsbuild/pants/blob/eed447a2971a417faca9b5744c8f38b7387e8604/src/python/pants/base/exception_sink.py#L545-L546

I don't think there's a string reason for us doing that, outside of our convention of often defining class vars as `None`. Were to instead properly set the default in the class definition, we could greatly simplify the type hints here and in `ExceptionSink`.

I'm going to try changing the default in this PR. I will wait to merge until you give explicit approval that the commit is safe.